### PR TITLE
ICU-21123 Support unit inflections in ICU4C

### PR DESCRIPTION
**Pull request by hugovdm**
Closes: icu-units#111

Progress:
- [x] inflect built-in units
- [x] support built-in unit gender
- [x] inflect builtin-per-builtin compound units
- [x] support builtin-per-builtin compound unit gender
- [x] inflect mixed units - CLDR-14502 filed to find out if I'm inflecting correctly
- N/A support mixed unit gender - ICU-21494
- icu-units#138: correctly implement Lateral Inheritance (current implementation is a shortcut adequate for CLDR 39 data)
- icu-units#139: implement deriveComponent plural per rules (currently hard-coded, matches all CLDR 39 rules)
- icu-units#140: support gender when not using unit-width-full-name
- icu-units#141: support gender for built-in compound units (treat as not-builtin to get the builtin-per-builtin gender)

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21123
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

**Comment by hugovdm** [comment]
Java failure surely unrelated to this PR:

    [junit] Running com.ibm.icu.dev.test.util.Trie2Test
    [junit] Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.203 sec
    [junit] TEST com.ibm.icu.dev.test.util.Trie2Test FAILED

File: icu4c/source/i18n/number_longnames.cpp [hunk 1]
```diff
@@ -169,6 +205,23 @@ void getMeasureData(const Locale &locale, const MeasureUnit &unit, const UNumber
...
     key.append(subtypeForResource, status);
 
+    // Grab desired case first, if available. Then grab nominative case to fill
+    // in the gaps.
+    //
+    // TODO(icu-units#138): implement spec-compliant fallback: case should
+    // fallback to "nominative" before falling back to no case. For the time
+    // being: as of CLDR 39 explicit "nominative" is not yet in use.
+    if (width == UNUM_UNIT_WIDTH_FULL_NAME && !unitDisplayCase.empty() &&
+        unitDisplayCase.compare("nominative") != 0) {
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:215]: shall we just process the nominative, even the process is doing nothing

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:215]: True! Thanks for the catch. That would even be more spec compliant. Here is the specification of the fallback process I should have implemented:
https://www.unicode.org/reports/tr35/tr35.html#Lateral_Inheritance
It should fall back to "nominative" before it falls back to no-case.

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp]: Done.

File: icu4c/source/i18n/unicode/numberformatter.h [hunk 2]
```diff
@@ -1131,30 +1131,30 @@ class U_I18N_API Scale : public UMemory {
...
 namespace impl {
 
-// Do not enclose entire Usage with #ifndef U_HIDE_INTERNAL_API, needed for a protected field
+// Do not enclose entire StringProp with #ifndef U_HIDE_INTERNAL_API, needed for a protected field
 /**
  * Manages NumberFormatterSettings::usage()'s char* instance on the heap.
  * @internal
  */
-class U_I18N_API Usage : public UMemory {
+class U_I18N_API StringProp : public UMemory {
```
Comment (younies) [inline icu4c/source/i18n/unicode/numberformatter.h:1139]: I still feel that `Usage` is more clear than StringProp

Comment (younies) [inline icu4c/source/i18n/unicode/numberformatter.h:1139]: I think also `StringProp` is better to be moved to Utilities

Comment (hugovdm) [reply icu4c/source/i18n/unicode/numberformatter.h:1139]: Usage is the wrong type for unitDisplayCase. This would be very confusing:

    Usage unitDisplayCase;

It isn't a usage. So what is it? It's really just a string container, for a string property of MacroProps.

Comment (hugovdm) [reply icu4c/source/i18n/unicode/numberformatter.h:1139]: This is NumberFormatter specific. E.g. friend classes:

    friend class NumberFormatterImpl;

It's a class for a MacroProps property, just like SymbolsWrapper, Grouper, and Padder, so I believe it belongs where those classes are found.

File: icu4c/source/i18n/number_longnames.cpp [hunk 3]
```diff
@@ -232,15 +282,96 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
     return UnicodeString(ptr, len);
 }
 
+// Load data originating from a deriveComponent CLDR element. The feature and
+// structure parameters must be null-terminated.
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:286]: Use function/Documentation  comment, /** */

File: icu4c/source/i18n/number_longnames.cpp [hunk 4]
```diff
@@ -169,6 +205,20 @@ void getMeasureData(const Locale &locale, const MeasureUnit &unit, const UNumber
...
+    // in the gaps.
+    //
+    // TODO(icu-units#138): check that fallback is spec-compliant
+    if (width == UNUM_UNIT_WIDTH_FULL_NAME && !unitDisplayCase.empty()) {
+        CharString caseKey;
+        caseKey.append(key, status);
+        caseKey.append("/case/", status);
+        caseKey.append(unitDisplayCase, status);
+
+        UErrorCode localStatus = U_ZERO_ERROR;
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:218]: I'm going with expanded TODOs at this point, letting issue 138 track this.

File: icu4c/source/i18n/number_longnames.cpp [hunk 5]
```diff
@@ -232,15 +282,96 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
...
 
+// Load data originating from a deriveComponent CLDR element. The feature and
+// structure parameters must be null-terminated.
+void getDeriveComponentRule(Locale locale,
+                            const char *feature,
+                            const char *structure,
+                            UnicodeString &value0,
+                            UnicodeString &value1,
+                            UErrorCode &status) {
+    StackUResourceBundle derivationsBundle, stackBundle;
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:293]: I really like to separate the loading data methods from the class, we can implement those functions part of `UnitsData` or maybe `InflectionUnitsData`.

Also, it can be in a different PR or in the same PR

but having loading data in the same class make the class more complicated.

File: icu4c/source/i18n/number_longnames.cpp [hunk 6]
```diff
@@ -232,15 +282,96 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
     return UnicodeString(ptr, len);
 }
 
+// Load data originating from a deriveComponent CLDR element. The feature and
+// structure parameters must be null-terminated.
+void getDeriveComponentRule(Locale locale,
+                            const char *feature,
+                            const char *structure,
+                            UnicodeString &value0,
+                            UnicodeString &value1,
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:291]: I prefer putting those returned value on struct such as `struct CLDRDerivedComponent`

Thus will help us in the future to add more values.

File: icu4c/source/i18n/number_longnames.cpp [hunk 7]
```diff
@@ -232,15 +282,96 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
...
+    }
+    ures_getByKey(stackBundle.getAlias(), "component", stackBundle.getAlias(), &status);
+    ures_getByKey(stackBundle.getAlias(), feature, stackBundle.getAlias(), &status);
+    ures_getByKey(stackBundle.getAlias(), structure, stackBundle.getAlias(), &status);
+    value0 = ures_getUnicodeStringByIndex(stackBundle.getAlias(), 0, &status);
+    value1 = ures_getUnicodeStringByIndex(stackBundle.getAlias(), 1, &status);
+}
+
+UnicodeString
+getDeriveCompoundRule(Locale locale, const char *feature, const char *structure, UErrorCode &status) {
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:318]: As I mentioned before, I prefer to add those kind of function an a separate class.

File: icu4c/source/i18n/number_longnames.cpp [hunk 8]
```diff
@@ -232,15 +282,96 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
...
+    ures_getByKey(stackBundle.getAlias(), feature, stackBundle.getAlias(), &status);
+    return ures_getUnicodeStringByKey(stackBundle.getAlias(), structure, &status);
+}
+
 ////////////////////////
 /// END DATA LOADING ///
 ////////////////////////
 
+// TODO: promote this somewhere? It's based on patternprops.cpp' trimWhitespace
+const UChar *trimSpaceChars(const UChar *s, int32_t &length) {
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:346]: This also part of `UnitsDataUtils` or other utils

File: icu4c/source/i18n/number_longnames.cpp (hunk 3 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:286]: Going with the "///" style, also known by doxygen. (But I suspect doxygen won't be generating docs for this function.)

File: icu4c/source/i18n/number_longnames.cpp (hunk 5 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:293]: Refactoring 3 year old code is out of scope for this PR, so I'm leaving this as is.

You're commenting on a function here, "void getDeriveComponentRule(...)" - so I don't know what class you're referring to as becoming "more complicated". And LongNameHandler itself exists precisely for the sake of loading and using units data.

File: icu4c/source/i18n/number_longnames.cpp [hunk 9]
```diff
@@ -232,15 +282,96 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
...
+    // TODO: use standard normal locale resolution algorithms rather than just grabbing language:
+    ures_getByKey(derivationsBundle.getAlias(), locale.getLanguage(), stackBundle.getAlias(), &status);
+    // TODO:
+    // - code currently assumes if the locale exists, the rules are there -
+    //   instead of falling back to root when the requested rule is missing.
+    // - investigate ures.h functions, see if one that uses res_findResource()
+    //   might be better (or use res_findResource directly), or maybe help
+    //   improve ures documentation to guide function selection?
+    if (status == U_MISSING_RESOURCE_ERROR) {
+        status = U_ZERO_ERROR;
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:307]: Done.

That tooling sounds interesting! Do you know where I can look, or who to ask, to learn a bit more?

File: icu4c/source/i18n/number_microprops.h [hunk 10]
```diff
@@ -82,6 +82,7 @@ struct MicroProps : public MicroPropsGenerator {
     UNumberDecimalSeparatorDisplay decimal;
     bool useCurrency;
     char nsName[9];
+    const char *gender;
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_microprops.h:85]: Done.

File: icu4c/source/i18n/number_longnames.cpp [hunk 11]
```diff
@@ -137,12 +165,19 @@ class PluralTableSink : public ResourceSink {
  * UNUM_UNIT_WIDTH_NARROW), or just *unitsShort* (for width
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:165]: Done.

File: icu4c/source/i18n/number_longnames.cpp [hunk 12]
```diff
@@ -295,17 +440,47 @@ void LongNameHandler::forCompoundUnit(const Locale &loc, const MeasureUnit &unit
         status = U_INTERNAL_PROGRAM_ERROR;
         return;
     }
+
+    StringPiece primaryCase, secondaryCase;
+    CharString primaryCase_, secondaryCase_;
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:453]: use the `variable_` for internal class variables, I kind feel it is confusing here.

File: icu4c/source/i18n/number_longnames.cpp (hunk 8 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:354]: I've added the TODO to figure out where to promote it to in the future. I think PatternProps (patternprops.h / patternprops.cpp) is a good place for it - but might require a bit of thought or discussion to find out what it should be called when it's based around u_isJavaSpaceChar.

 (There is no UnitsDataUtils in ICU4C.)

File: icu4c/source/i18n/number_longnames.cpp (hunk 12 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:453]: Thanks for catching! Agreed - bad names. I've renamed them to *Mem to indicate they're in charge of holding memory. 

FWIW I took inspiration from this one:
```
static const MeasureUnitImpl& MeasureUnitImpl::forMeasureUnit(
        const MeasureUnit& measureUnit, MeasureUnitImpl& memory, UErrorCode& status);
```

File: icu4c/source/i18n/number_longnames.cpp [hunk 13]
```diff
@@ -295,17 +440,47 @@ void LongNameHandler::forCompoundUnit(const Locale &loc, const MeasureUnit &unit
...
+            } else {
+                secondaryCase_.appendInvariantChars(uVal1, status);
+                secondaryCase = secondaryCase_.toStringPiece();
+            }
+        } else {
+            // Ignore failures in production code, assert-fail in debug mode.
+            U_ASSERT(false);
+            status = U_ZERO_ERROR;
+        }
+    }
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:476]: nit/ many nested if statements

File: icu4c/source/i18n/number_longnames.cpp (hunk 6 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:293]: I guess that is necessary in Java, because Java doesn't have output parameters - in Java one must either modify an object passed as a parameter, or return an object that can encapsulate the data in question.

I looked over some ICU4C and ICU4J code to try to get a sense of what's idiomatic, but maybe @sffc is quicker in advising here: I hope with only two strings, having value0 and value1 parameters is idiomatic enough in C++ - I don't yet know what's idiomatic in ICU4J.

One random sample I found: SpoofChecker results:
https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/text/SpoofChecker.CheckResult.html
In C++:
https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/uspoof_8h.html
It's quite a bit more complicated than just two strings though, and I don't think something like that is worth the overhead for a function as simple as getDeriveComponentRule.

File: icu4c/source/i18n/number_longnames.cpp [hunk 14]
```diff
@@ -350,7 +547,7 @@ UnicodeString LongNameHandler::getUnitDisplayName(
         return ICU_Utility::makeBogusString();
     }
     UnicodeString simpleFormats[ARRAY_LENGTH];
-    getMeasureData(loc, unit, width, simpleFormats, status);
+    getMeasureData(loc, unit, width, "", simpleFormats, status);
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:558]: discuss this offline

File: icu4c/source/i18n/number_longnames.cpp [hunk 15]
```diff
@@ -364,7 +561,7 @@ UnicodeString LongNameHandler::getUnitPattern(
         return ICU_Utility::makeBogusString();
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:569]: why there is no getUnitPattern in Java. ?

File: icu4c/source/i18n/number_longnames.cpp (hunk 7 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:326]: In Java: separate class, separate file (or nested class). In C++: this is just a function. The .cpp files tend to get rather big sometimes - but that's just file length, not really complexity?

The LongNameHandler java class has:
- PluralTableSink (subclass)
- getMeasureData
- getCurrencyLongNameData
- getPerUnitFormat

This class looks like a good place for getDeriveCompoundRule to me. Component and compound rules are all related to producing names for units.

File: icu4c/source/i18n/number_longnames.cpp (hunk 15 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:569]: I see there's quite a big difference in MeasureFormat.formatMeasure implementations - which is the only place where getUnitPattern is used. MeasureFormat is old / not recommended, so I think you can ignore its absence.

I documented the C++ one to stop myself from worrying about it. :-)

File: icu4c/source/i18n/number_longnames.cpp (hunk 14 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:558]: Happy to discuss!
(As to this parameter: display names aren't inflected.)

File: icu4c/source/i18n/unicode/numberformatter.h [hunk 16]
```diff
@@ -2658,6 +2682,13 @@ class U_I18N_API FormattedNumber : public UMemory, public FormattedValue {
      * @draft ICU 68
      */
     MeasureUnit getOutputUnit(UErrorCode& status) const;
+
+    /**
+     * The gender of the formatted output.
+     *
```
Comment (younies) [inline icu4c/source/i18n/unicode/numberformatter.h:2688]: start the comment with verb in form "starts, gets ... etc)

File: icu4c/source/test/intltest/numbertest.h [hunk 17]
```diff
@@ -64,6 +64,8 @@ class NumberFormatterApiTest : public IntlTestWithFieldPosition {
     void unitUsageErrorCodes();
     void unitUsageSkeletons();
     void unitCurrency();
+    void unitInflections();
+    void unitGender();
```
Comment (younies) [inline icu4c/source/test/intltest/numbertest.h:68]: +1

File: icu4c/source/i18n/unicode/numberformatter.h (hunk 16 above)
Comment (hugovdm) [reply icu4c/source/i18n/unicode/numberformatter.h:2689]: Done.

File: icu4c/source/i18n/number_longnames.cpp (hunk 13 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:476]: Ooh, I like this idea:
I could refactor "getDeriveComponentRule" in a way that produces primaryCase and secondaryCase instead of producing uVal0 and uVal1.

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:476]: Main problem: needing to take care of storing the converted strings means I would have to pass in primaryCaseMem, secondaryCaseMem, primaryCase, and secondaryCase. Might actually make sense to make a little class out of this.

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:476]: Conceptually:

    DerivedComponents derivedCases(loc, "case", "per");
    primaryCase = derivedCases.value0;
    secondaryCase = derivedCases.value1;

Means I can reuse it for "times", "power" and "prefix, as well as using it for both "case" and "plural"!

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp]: Done! PTAL.

I'm not doing the same thing for deriveCompound because it does not look like it has much benefit: all it could do is in any case return a 0 or 1. (Could be reconsidered once I'm using `getDeriveCompoundRule` more.)

File: icu4c/source/i18n/number_longnames.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/i18n/number_longnames.cpp:215]: +1

File: icu4c/source/i18n/number_longnames.cpp (hunk 15 above)
Comment (younies) [reply icu4c/source/i18n/number_longnames.cpp:600]: +1

**Review by younies — DISMISSED** [review]

File: icu4c/source/i18n/number_longnames.h [hunk 18]
```diff
@@ -205,8 +229,11 @@ class LongNameMultiplexer : public MicroPropsGenerator, public UMemory {
     // `units`. An individual unit might be a mixed unit.
     static LongNameMultiplexer *forMeasureUnits(const Locale &loc,
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.h:230]: comments in Java
```
    /**
     * Construct a localized MixedUnitLongNameHandler for the specified
     * MeasureUnit. It must be a MIXED unit.
     * <p>
     *
     * @param locale    The desired locale.
     * @param mixedUnit The mixed measure unit to construct a
     *                  MixedUnitLongNameHandler for.
     * @param width     Specifies the desired unit rendering.
     * @param unitDisplayName
     * @param rules     PluralRules instance.
     * @param parent    MicroPropsGenerator instance.
     */
```

File: icu4c/source/i18n/number_longnames.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/number_longnames.cpp]: gdb and other debuggers allow watching variables for changes.

**Review by sffc — DISMISSED** [review]

**Review by younies — APPROVED** [review]

File: icu4c/source/i18n/number_longnames.h (hunk 18 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.h:230]: FYI: what you pasted here is the MixedUnitLongNameHandler, where you pasted it was the LongNameMultiplexer.

File: icu4c/source/i18n/number_longnames.cpp (hunk 8 above)
Comment (younies) [reply icu4c/source/i18n/number_longnames.cpp:408]: +1

File: icu4c/source/i18n/number_longnames.cpp [hunk 19]
```diff
@@ -232,15 +284,156 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
...
+};
+
+UnicodeString
+getDeriveCompoundRule(Locale locale, const char *feature, const char *structure, UErrorCode &status) {
+    StackUResourceBundle derivationsBundle, stackBundle;
+    ures_openDirectFillIn(derivationsBundle.getAlias(), NULL, "grammaticalFeatures", &status);
+    ures_getByKey(derivationsBundle.getAlias(), "grammaticalData", derivationsBundle.getAlias(),
+                  &status);
+    ures_getByKey(derivationsBundle.getAlias(), "derivations", derivationsBundle.getAlias(), &status);
+    // TODO: use standard normal locale resolution algorithms rather than just grabbing language:
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:386]: period

Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:386]: I made it a colon to indicate I'm talking about the next line. That line contains "locale.getLanguage()", "just grabbing language".

File: icu4c/source/i18n/number_longnames.cpp [hunk 20]
```diff
@@ -295,17 +502,24 @@ void LongNameHandler::forCompoundUnit(const Locale &loc, const MeasureUnit &unit
...
+    getMeasureData(loc, perUnit, width, derivedPerCases.value1(), secondaryData, status);
     if (U_FAILURE(status)) {
         return;
     }
 
+    // TODO(icu-units#139): implement these rules:
+    // <deriveComponent feature="plural" structure="per" ...>
+    // This has impact on multiSimpleFormatsToModifiers(...) below too.
+    // These rules are currently (ICU 69) all the same and hard-coded below.
     UnicodeString perUnitFormat;
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:523]: Nit:
```
        // TODO(icu-units#139): implement these rules:
        //    - <deriveComponent feature="plural" structure="per" ...>
        //    - This has impact on multiSimpleFormatsToModifiers(...) below too.
        // 
        // These rules are currently (ICU 69) all the same and hard-coded below.
```

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 21]
```diff
@@ -1926,6 +1928,197 @@ void NumberFormatterApiTest::unitCurrency() {
             u"123,12 CN¥");
 }
 
+void NumberFormatterApiTest::runUnitInflectionsTestCases(UnlocalizedNumberFormatter unf,
+                                                         const UChar *skeleton,
+                                                         const UChar *conciseSkeleton,
+                                                         const UnitInflectionTestCase *cases,
```
Comment (younies) [inline icu4c/source/test/intltest/numbertest_api.cpp:1934]: I prefer if it takes only one test case at a time.

Comment (hugovdm) [reply icu4c/source/test/intltest/numbertest_api.cpp:1934]: In C++ array construction can be done nice and concisely - e.g. in my latest PR my tests are looking like this:

        const UnitInflectionTestCase meterCases[] = {
            // Check up on the basic values that the compound patterns below are
            // derived from:
            {"meter", "de", nullptr, 1, u"1 Meter"},
            {"meter", "de", "genitive", 1, u"1 Meters"},
            {"meter", "de", nullptr, 2, u"2 Meter"},
            {"meter", "de", "dative", 2, u"2 Metern"},
            {"mile", "de", nullptr, 1, u"1 Meile"},
            {"mile", "de", nullptr, 2, u"2 Meilen"},
            {"day", "de", nullptr, 1, u"1 Tag"},
            {"day", "de", "genitive", 1, u"1 Tages"},
            {"day", "de", nullptr, 2, u"2 Tage"},
            {"day", "de", "dative", 2, u"2 Tagen"},
            {"decade", "de", nullptr, 1, u"1\u00A0Jahrzehnt"},
            {"decade", "de", nullptr, 2, u"2\u00A0Jahrzehnte"},
        ...

In Java I believe there'd always be a need to place the "TestCase()" constructor around this, for example. In which case we might as well have `runTestCase("decade", "de", nullptr, 2, u"2\u00A0Jahrzehnte")` in Java: thus I'm agreeing it to be reasonable to diverge from C++ in this.

File: icu4c/source/i18n/number_longnames.cpp (hunk 20 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:523]: This is still a TODO, even after my next PR, because it does not impact ICU 69 with CLDR 39.
